torch/csrc/jit/passes/peephole_list_idioms.cpp

```diff
@@ -21,17 +22,312 @@ c10::optional<size_t> normalizeIndex(int64_t index, size_t len) {
+          changed_ = true;
+          WithInsertPoint guard(n);
+          n->output()->replaceAllUsesWith(
+              graph_->insertConstant(static_cast<int64_t>(*maybe_len)));
+        }
+      }
+
+      if (n->kind() == prim::If) {
+        IfView if_n(n);
+        bool has_cond_ref = info_.count(if_n.cond()) != 0;
+        ListRefinement empty;
+        auto true_block_refinements = RefineListLens(
+            if_n.thenBlock(),
+            has_cond_ref ? info_[if_n.cond()].true_refine() : empty);
+        auto false_block_refinements = RefineListLens(
+            if_n.elseBlock(),
+            has_cond_ref ? info_[if_n.cond()].false_refine() : empty);
+        bool true_block_throws = throwing_blocks_.count(if_n.thenBlock());
+        bool false_block_throws = throwing_blocks_.count(if_n.elseBlock());
+
+        // if one block throws, the refinements for the other block
+        // become present in the current block, and all bool outputs
+        // of the if node take their refinements from non throwing block
+        // output
+
+        if (true_block_throws || false_block_throws) {
+          if (true_block_throws && false_block_throws) {
+            throwing_blocks_.insert(b);
+            continue;
+          }
+          if (true_block_throws) {
+            block_refinements.insert(
+                false_block_refinements.begin(), false_block_refinements.end());
+          } else {
+            block_refinements.insert(
+                true_block_refinements.begin(), true_block_refinements.end());
+          }
+          Block* non_throwing_block =
+              true_block_throws ? n->blocks().at(1) : n->blocks().at(0);
+          for (size_t i = 0; i < if_n.outputs().size(); ++i) {
+            if (info_.count(non_throwing_block->outputs().at(i))) {
+              info_[n->outputs().at(i)] =
+                  info_[non_throwing_block->outputs().at(i)];
+            }
+          }
+          continue;
+        }
+
+        // if either block has a constant bool output, e.g. `true` on the
+        // truee block, then for the `false` value we can take the false
+        // refinements from the other block and from the other block value bc
+        // if the output is false it had to have come from the false block.
+        // Otherwise, just take intersection of refinements
+
+        for (size_t i = 0; i < if_n.outputs().size(); ++i) {
+          if (!(if_n.outputs().at(i)->type() == BoolType::get())) {
+            continue;
+          }
+          Value* true_v = if_n.thenOutputs().at(i);
+          Value* false_v = if_n.elseOutputs().at(i);
+
+          if (!info_.count(true_v) && !info_.count(false_v)) {
+            continue;
+          }
+
+          BoolRefinements out;
+          if (auto maybe_bool = constant_as<bool>(true_v)) {
+            if (*maybe_bool) {
+              out = BoolRefinements::FalseRefinements(unionRefinements(
+                  info_[false_v].false_refine(), false_block_refinements));
```

Comment:
From Line 258-259, it indicates `info_count[true_v] || info_count[false_v]`, so does `info_[false_v]` always exists under this condition? If so, please add some comments because it is not obvious.

Reply:
Yea, will do. We never associate refinements with constant values so yes that is the invariant.